includes/class-wc-order-product-lookup-helper.php

```diff
@@ -0,0 +1,207 @@
+		add_action( 'woocommerce_delete_product_variation', array( __CLASS__, 'handle_delete_product' ), 10, 1 );
+		add_action( 'woocommerce_after_order_object_save', array( __CLASS__, 'handle_new_order' ), 10, 1 );
+		add_action( 'woocommerce_update_order', array( __CLASS__, 'handle_update_order' ), 10, 1 );
+	}
+
+	/**
+	 * Handler for deleting an order, keep the lookup table in sync.
+	 *
+	 * @param int $order_id Order ID.
+	 *
+	 * @since 3.3.0
+	 */
+	public static function handle_delete_order( $order_id ) {
+		global $wpdb;
+
+		$wpdb->delete(
+			$wpdb->prefix . 'woocommerce_order_product_lookup',
+			array(
+				'order_id' => $order_id,
+			),
+			array(
+				'%d',
+			)
+		); // WPCS: cache ok.
+	}
+
+	/**
+	 * Handler for deleting a product, keep the lookup table in sync.
+	 *
+	 * @param int $product_id Product ID.
+	 *
+	 * @since 3.3.0
+	 */
+	public static function handle_delete_product( $product_id ) {
+		global $wpdb;
+
+		$wpdb->delete(
+			$wpdb->prefix . 'woocommerce_order_product_lookup',
+			array(
+				'product_id' => $product_id,
+			),
+			array(
+				'%d',
+			)
+		); // WPCS: cache ok.
+	}
+
+	/**
+	 * Handler for saving an order, keep the lookup table in sync.
+	 *
+	 * @param WC_Order $order Order object.
+	 *
+	 * @since 3.3.0
+	 */
+	public static function handle_new_order( $order ) {
+		global $wpdb;
+
+		$order_items = $order->get_items();
+
+		foreach ( $order_items as $item_id => $item ) {
+			$item_product_id = $item->get_product_id();
+
+			// Skip if this line item doesn't have a product id associated.
+			if ( empty( $item_product_id ) ) {
+				continue;
+			}
+
+			$data = array(
+				'user_email'           => $order->get_billing_email(),
+				'order_id'             => $order->get_id(),
+				'user_id'              => $order->get_customer_id(),
+				'product_id'           => $item_product_id,
+				'parent_order_item_id' => $item_id,
+			);
+
+			$wpdb->insert(
+				$wpdb->prefix . 'woocommerce_order_product_lookup',
+				$data,
+				array(
+					'%s',
+					'%d',
+					'%d',
+					'%d',
+					'%d',
+				)
+			);
+		}
+	}
+
+	/**
+	 * Handler for updating an order, keep the lookup table in sync.
+	 *
+	 * @param int $order_id Order ID.
+	 *
+	 * @since 3.3.0
+	 */
+	public static function handle_update_order( $order_id ) {
+		// Make sure we delete the old order from the lookup database, as it may have its line items updated.
+		self::handle_delete_order( $order_id );
+
+		$order = wc_get_order( $order_id );
+
+		if ( is_a( $order, 'WC_Order' ) ) {
+			self::handle_new_order( $order );
+		}
+	}
+
+	/**
+	 * Check if a product is purchased, given either email or user id.
+	 *
+	 * @param  int        $product_id       Product ID.
+	 * @param  string|int $email_or_user_id Either email or user id.
+	 * @return bool       Whether product is purchased or not.
+	 *
+	 * @since 3.3.0
+	 */
+	public static function is_product_purchased( $product_id, $email_or_user_id ) {
```

Comment:
With some extra effort this could be rewritten to handle arrays of products and:

* Check if any product from an array of products has been purchased.
* Check if all products in an array of products have been purchased.

Would require an extra `$relationship` argument.

Disregarding the slowness of `wc_customer_bought_product`, its main limitation has always been its inability to handle arrays of products, although its main query returned an array of IDs.